editoast/src/modelsv2/projects.rs

```diff
@@ -0,0 +1,265 @@
+use crate::views::pagination::{Paginate, PaginatedResponse};
+use crate::views::projects::{ProjectError, ProjectWithStudies};
+use crate::DbPool;
+use actix_web::web::Data;
+use async_trait::async_trait;
+use chrono::{NaiveDateTime, Utc};
+use diesel::{sql_query, ExpressionMethods, QueryDsl};
+use diesel_async::{AsyncPgConnection as PgConnection, RunQueryDsl};
+use editoast_derive::ModelV2;
+use serde::{Deserialize, Serialize};
+use utoipa::ToSchema;
+
+use crate::models::List;
+
+crate::schemas! {
+    Ordering,
+    Project,
+    Tags,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, ModelV2, ToSchema, PartialEq)]
+#[model(table = crate::tables::project)]
+pub struct Project {
+    pub id: i64,
+    pub name: String,
+    pub objectives: String,
+    pub description: String,
+    pub funders: String,
+    pub budget: i32,
+    pub creation_date: NaiveDateTime,
+    pub last_modification: NaiveDateTime,
+    #[model(remote = "Vec<Option<String>>")]
+    pub tags: Tags,
+    #[model(column = "image_id")]
+    pub image: Option<i64>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default, ToSchema, PartialEq)]
+pub struct Tags(Vec<String>);
+
+impl Tags {
+    pub fn new(value: Vec<String>) -> Self {
+        Self(value)
+    }
+}
+
+impl From<Vec<Option<String>>> for Tags {
+    fn from(value: Vec<Option<String>>) -> Self {
+        Self(value.into_iter().flatten().collect())
+    }
+}
+
+impl From<Tags> for Vec<Option<String>> {
+    fn from(value: Tags) -> Self {
+        value.0.into_iter().map(Some).collect()
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, Default, ToSchema)]
+pub enum Ordering {
+    NameAsc,
+    NameDesc,
+    CreationDateAsc,
+    CreationDateDesc,
+    #[default]
+    LastModifiedDesc,
+    LastModifiedAsc,
+}
+
+impl Ordering {
+    pub fn to_sql(&self) -> &str {
+        match *self {
+            Ordering::NameAsc => "LOWER(t.name) ASC",
+            Ordering::NameDesc => " LOWER(t.name) DESC",
+            Ordering::CreationDateAsc => "creation_date",
+            Ordering::CreationDateDesc => "creation_date DESC",
+            Ordering::LastModifiedAsc => "last_modification",
+            Ordering::LastModifiedDesc => "last_modification DESC",
+        }
+    }
+}
+
+impl Project {
+    /// This function takes a filled project and update to now the last_modification field
+    pub async fn update_last_modified(self, db_pool: Data<DbPool>) -> Result<Option<Project>> {
+        let mut conn = db_pool.get().await?;
+        self.update_last_modified_conn(&mut conn).await
+    }
+
+    pub async fn update_last_modified_conn(
+        mut self,
+        conn: &mut PgConnection,
+    ) -> Result<Option<Project>> {
+        self.last_modification = Utc::now().naive_utc();
+        self.save(conn).await?;
+        Ok(Some(self))
+    }
+
+    pub async fn with_studies(self, db_pool: Data<DbPool>) -> Result<ProjectWithStudies> {
```

Comment:
```suggestion
    pub async fn with_studies(self, conn: &mut diesel_async::...) -> Result<usize> {
```

That way you keep `ProjectWithStudies` in the `views` module. You can make a `ProjectWithStudies::new(project: Project)` function.

Reply:
if I understood correctly `with_studies` will just return studies count ? in this case, I think it's better to use `Study` struct directly instead of `Project`

Reply:
Yep, you can rename it `studies_count` btw. Using the `Study` to retrieve the count seems weird to me, because to get the count you need to start from the project and explore its foreign refs `project --> COUNT(studies)`. Using the study struct just adds another level of indirection to find the project `study --> [ project --> COUNT(studies) ]`, and that doesn't bring anything.

If it's a matter a namespaces, it's alright to let the db structure "leak" since we're in "`models`".

Reply:
Done.